openpathsampling/treelogic.py

```diff
@@ -1,169 +1,247 @@
-            return self._subnodes[item]
+        hashable object
+            the unique (hashable) key to identify each node
 
-        if type(item) is list:
-            # this is assumed to be a tree
-            if self._default_match(item[0], self):
-                if len(item) > 1:
-                    for ch in self._subnodes:
-                        r = ch[item[1]]
-                        if r is not None:
-                            return r
-                    return None
-                else:
-                    return self
-            else:
-                return None
+        Notes
+        -----
+        This is often specific to the node type and hence overridden by the
+        target tree
+        """
+        return hex(id(self))
 
+    @property
+    def _choices(self):
+        if self._node_type == self.NODE_TYPE_ALL:
+            return [[sub for sub in self._subnodes]]
+        elif self._node_type == self.NODE_TYPE_ACCUMULATE:
+            return [self._subnodes[:n+1] for n in range(0, len(self._subnodes))]
+        elif self._node_type == self.NODE_TYPE_ONE:
+            leaves = [[sub] for sub in self._subnodes]
+
+            unique_leaves = list(set([ tuple(l) for l in leaves ]))
+
+            return unique_leaves
+        elif self._node_type == self.NODE_TYPE_POWER:
+            s = list(self._subnodes)
+            return itertools.chain.from_iterable(
+                itertools.combinations(s, r) for r in range(len(s)+1)
+            )
+        elif self._node_type == self.NODE_TYPE_NONE:
+            return []
+        elif hasattr(self._node_type, '__call__'):
+            return self._node_type()
+        else:
+            raise RuntimeError('Node has no type !')
 
-    def __reversed__(self):
-        """
-        Traverse the whole tree in post-order
+    def treeprint(self):
+        return str(self.head) + "\n" + TreeSetMixin._indent("\n".join(map(lambda x : x.treeprint(), self.branches)))
 
-        Returns
-        -------
-        Iterator
-            an iterator that traverses the tree in post-order
-        """
-        for subchange in self._subnodes:
-            for change in reversed(subchange):
-                yield change
+    def locate(self, item):
+        l = [key for key, value in self.locators().iteritems() if self._default_match(value, item)]
+        if len(l) == 0:
+            return None
+        elif len(l) == 1:
+            return l[0]
+        else:
+            return l
+
+    def pick(self, item):
+        loc = self.locate(item)
+        if loc is None:
+            return loc
+        elif type(loc) is list:
+            return [self[l] for l in loc]
+        else:
+            return self[loc]
 
-        yield self
+    @property
+    def is_sequential(self):
+        return len(self._choices) < 2
 
-    def __len__(self):
-        """
-        Returns the total number of Changes mad in a move.
+    @property
+    def deterministic(self):
```

Comment:
I think I should first add better documentation, but for now deterministic is `True` if the sequence of submoves is fixed in the sense that there is only one way that the movers below will be called. E.g. a `RandomChoiceMover` is not deterministic, but a `SequentialMover` is. This is used to check for the minimal representation since everything that is deterministic does not need to be part of the minimal subtree (besides maybe the first mover to check if the deterministic sequence is called or not).